### PR TITLE
[sentry-native] Support for builds without a backend.

### LINKED ISSUE #27696: [sentry-native] support different backends via feature flags
****

**Is your feature request related to a problem? Please describe.**

I only use sentry for generic event reporting, and don't need the crashpad machinery.

**Proposed solution**
Add feature flags for different backends, including "none" backend.

**Describe alternatives you've considered**
Compile myself, but its a headache as its using cmake where as my projects use msbuild.

### PR DESCRIPTION
**Pull request by FrankXie05**
**Describe the pull request**

- #### What does your PR fix?
  Fixes https://github.com/microsoft/vcpkg/issues/27696
  Note: Officially supports builds without a backend and adds the warning.

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
How do you set these defines at vcpkg install time? Or does vcpkg inherit the defines from the parent project at msbuild time?

**Comment by FrankXie05** [comment]
@AudriusButkevicius I changed it to a feature. Add parameter command `./vcpkg install sentry-native[none]` to use it. :)

File: ports/sentry-native/vcpkg.json [hunk 1]
```diff
@@ -19,5 +20,10 @@
       "host": true
     },
     "zlib"
-  ]
+  ],
+  "features": {
+    "none": {
+      "description": "This builds sentry-native without a backend, so it does not handle crashes at all. It is primarily used for tests."
```
Comment (AudriusButkevicius) [inline ports/sentry-native/vcpkg.json:26]: I think the "used for tests" is wrong. Sentry is useful for other things than just tests without a backend, as I explained on my original ticket.

Comment (FrankXie05) [reply ports/sentry-native/vcpkg.json:26]: I just copied the instructions from upstream. 🧐  https://github.com/getsentry/sentry-native#compile-time-options
````
none: This builds sentry-native without a backend, so it does not handle crashes at all. It is primarily used for tests.
````

Comment (AudriusButkevicius) [reply ports/sentry-native/vcpkg.json:26]: Sadness :/

I guess I won't argue with the people making the software, even if I disagree.

Comment (FrankXie05) [reply ports/sentry-native/vcpkg.json:26]: Sorry, unless there is a good reason, I will still take the official statement. 😔

**Review by AudriusButkevicius — DISMISSED** [review]

**Review by dg0yt — CHANGES_REQUESTED** [review]
The feature `none` is invalid because it is not additive.

What could be done is implement an opt-out feature `backend` or `default-backend` which enables the default backend for the given platform.
What must not be done is implementing alternatives via features.

**Comment by FrankXie05** [comment]
For port `sentry-native`, any platform has a default backend, but it also supports `none` (except IOS).
https://github.com/getsentry/sentry-native#compile-time-options
image

**Review by JonLiu1993 — DISMISSED** [review]

**Comment by FrankXie05** [comment]
The feature `none` has been tested successfully locally.

**Comment by dg0yt** [comment]
> The feature `none` is invalid because it is not additive.
> 
> What could be done is implement an opt-out feature `backend` or `default-backend` which enables the default backend for the given platform. What must not be done is implementing alternatives via features.

**Please re-read immediately, remove approval/info:reviewed, and fix. **

**Comment by FrankXie05** [comment]
@dg0yt Sorry, I don't understand what you mean. What is `an opt-out feature backend` and what is its implementation logic?

**Comment by dg0yt** [comment]
Default features are installed unless the user explicitly requests to not install them (aka *opt-out*). Opt-out is by requesting the pseudo-feature `core` in classic mode or specifying `"default-features": false` in manifest mode, for that package. This opt-out is blocked by any other required port which (implicitly) depends on the port's default features.

In particular, default features shall include what users commonly expect from the package when not installed with vcpkg. For example the default backend.

*Features must be additive.* If a user or port doesn't opt out from default features and relies on the resulting (default) binary feature set, it must not break if an additional feature is added. The proposed feature `none` breaks this rule by taking away the default backend.

What can be done is to add a feature `backend`, listed in `default-features`, and add this to `portfile.cmake`:
~~~cmake
vcpkg_list(SET FEATURE_OPTIONS)
if(NOT "backend" IN_LIST FEATURES)
    vcpkg_list(APPEND FEATURE_OPTIONS -DSENTRY_BACKEND=none)
    message(STATUS "This build doesn't include a backend to handle crashes. It is primarily used for tests.")
endif()
~~~

You may even add per-platform dependencies to feature `backend`.

File: ports/sentry-native/portfile.cmake [hunk 2]
```diff
@@ -34,6 +41,7 @@ endif()
 vcpkg_cmake_configure(
     SOURCE_PATH "${SOURCE_PATH}"
     OPTIONS
+        ${FEATURE_OPTIONS}
         -DSENTRY_BUILD_TESTS=OFF
         -DSENTRY_BUILD_EXAMPLES=OFF
         -DSENTRY_BACKEND=${SENTRY_BACKEND}
```
Comment (dg0yt) [inline ports/sentry-native/portfile.cmake:47]: Note that my proposal was aiming at removing the explicit selection of the defautl `SENTRY_BACKEND` which is still left in lines 17-28 and used here, overwriting the effect of `FEATURE_OPTIONS`. 
If the package is smart enough to select a default backend, I don't see a need to duplicate this in the portfile.
If it needs this explicit selection, you will have to handle feature `backend` there.

**Review by dg0yt — CHANGES_REQUESTED** [review]

File: ports/sentry-native/vcpkg.json [hunk 3]
```diff
@@ -23,5 +23,4 @@
       "name": "zlib",
       "platform": "ios | osx | windows"
     }
-  ]
```
[the remark this answers is not in the document]
Comment (FrankXie05) [reply ports/sentry-native/vcpkg.json]: Yes, I missed this.

**Pull request closed without merging**